### ISSUE TITLE
How to find the best hyperparameters that will give me the best performance with my dataset in YOLOv8

### ISSUE THREAD

**Issue by ayadashash** [ultralytics/ultralytics#7749]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I'm working on YOLOv8 and I want to know if there is a way to find the best hyperparameters that will give me the best performance with my dataset. I've read about hyperparameters tuning but I didn't understand how It will work I also don't know which file that I should be adjusting in the respiratory of Yolov8. Changing parameters like optimizer, batch size and images size and making different combinations and testing the performance of the model is time and computational resources consuming. so I'm seeking for way to do this in more easy way. Thank you for your time and assistance.

**Comment by glenn-jocher** [comment]
@ayadashash hello! 👋

Finding the best hyperparameters for your dataset is indeed a crucial step to maximize the performance of your YOLOv8 model. To facilitate this, you can use the hyperparameter evolution feature available in YOLOv8. This feature automatically tests different combinations of hyperparameters over several training runs to find the best set for your specific dataset.

To start the hyperparameter evolution process, you'll need to use the `--evolve` flag when running your training command. This will initiate a search where the algorithm tries different hyperparameter values, evaluates the performance, and iteratively refines the parameters.

Keep in mind that this process can be computationally intensive and may take a significant amount of time, depending on the complexity of your dataset and the computational resources at your disposal. It's a trade-off between manual tuning and automated searching, but it can save you time in the long run by systematically exploring the hyperparameter space.

For more detailed guidance on how to use the hyperparameter evolution feature, please refer to our documentation at https://docs.ultralytics.com. Good luck with your model training! 🚀

**Comment by ayadashash** [comment]
@glenn-jocher Hii
first of all thank you so much for your fast response. 
I tried the evolution feature available in YOLOv8 which is Ray tune but it gives me error. I'm not sure actually if I used it probably. Please see below the code used for traning:
```python
from ray import tune
model = YOLO('yolov8n.pt')
result = model.tune(data='/content/drive/MyDrive/my_data.yaml', space={"epochs": tune.choice([50,60,70]),
                                                                                            "batch size": tune.choice([8,16,32]),
                                                                                            "imgsz":tune.choice([640,1280]),
                                                                                            "optimizer":tune.choice(['Adam','SGD','Adamax'])},use_ray=True)

``` 
I also have one more question, In YOLOv8 how can validate my model over the test data. I mean in YOLOv7 I used to run this to do so:
```python
! python test.py --data /content/drive/MyDrive/my_data.yaml --weights ayolo75_best.pt  --name aya_yolo7_12 --task test
``` 
How can I do the same thing with YOLOv8 is it by changing the mode to predict?

**Comment by glenn-jocher** [comment]
@ayadashash Hi there!

For the hyperparameter tuning error you're encountering with Ray Tune, it's important to ensure that you're following the YOLOv8 documentation for the correct usage of the tuning feature. The code snippet you've provided seems to be a mix of API calls that may not be directly applicable to YOLOv8's implementation. Please refer to our documentation on hyperparameter tuning for the proper setup and execution.

Regarding validation over test data in YOLOv8, you can use the `Val` mode to evaluate your model's performance. The command structure is similar to YOLOv7 but adapted for YOLOv8's CLI. You don't need to change the mode to `Predict` as that is used for making predictions on new data rather than evaluating against a test set.

For more information on how to validate your model with YOLOv8, please visit our documentation on the `Val` mode. Happy training! 🌟

**Comment by ayadashash** [comment]
thank you so much that was helpful 
I want to ask one more question, if Iam training yolo with dataset of one class, and I want to add background images (images with no object) I've seen that ultralytics recommends 0-10% of training images, is there any paper that confirm this ? and by adding background images should i make the number of classes 2 instead of 1 ?
thank you

**Comment by glenn-jocher** [comment]
@ayadashash Glad to hear the information was helpful!

The recommendation to include 0-10% of background images in your dataset is based on empirical evidence from our experience and community feedback rather than a specific paper. The decrease in mAP when increasing the number of background images could be due to the model becoming more conservative in predicting the object class, as it sees more examples of non-object (background) scenarios.

When training with background images, you should still keep the number of classes as 1 if you're only detecting one object class. The background does not count as a class; it's simply there to improve the robustness of your model to negative examples.

Keep experimenting with the ratio of background images to find the sweet spot for your particular dataset. 🎯

**Comment by ayadashash** [comment]
If I want to give the background images a class as 0 and the images that has object a class of 1 and take the output as .csv file, can I do this?

**Comment by glenn-jocher** [comment]
@ayadashash yes, you can label background images as class 0 and object images as class 1. To get the output as a .csv file, you can run inference with your model and then write the predictions to a CSV file using a custom script. Remember, the model's output format will need to be processed accordingly. 🛠️

**Comment by sxmair** [comment]
> @ayadashash hello! 👋
> 
> Finding the best hyperparameters for your dataset is indeed a crucial step to maximize the performance of your YOLOv8 model. To facilitate this, you can use the hyperparameter evolution feature available in YOLOv8. This feature automatically tests different combinations of hyperparameters over several training runs to find the best set for your specific dataset.
> 
> To start the hyperparameter evolution process, you'll need to use the `--evolve` flag when running your training command. This will initiate a search where the algorithm tries different hyperparameter values, evaluates the performance, and iteratively refines the parameters.
> 
> Keep in mind that this process can be computationally intensive and may take a significant amount of time, depending on the complexity of your dataset and the computational resources at your disposal. It's a trade-off between manual tuning and automated searching, but it can save you time in the long run by systematically exploring the hyperparameter space.
> 
> For more detailed guidance on how to use the hyperparameter evolution feature, please refer to our documentation at https://docs.ultralytics.com. Good luck with your model training! 🚀

May I know How do I utilise the evolve command in CLI? currently my command is as below but I want to try autotune hyperparameters to improve my model. 

yolo detect train data=data.yaml model=yolov8s.pt epochs=30 imgsz=640 device=2,3

**Comment by HaTrungTn23** [comment]
@glenn-jocher ​Hello, i try to use:
model = YOLO("/kaggle/working/yolov8n-cls.pt")
model.tune(data="/kaggle/input/dataset1/Data_300", epochs=30, iterations=3, plots=False, save=False, val=False)

and it return:

Command '['yolo', 'train', 'task=classify', 'mode=train', 'model=/kaggle/working/yolov8n-cls.pt', 'data=/kaggle/input/dataset1/Data_300', 'epochs=30', 'time=None', 'patience=100', 'batch=16', 'imgsz=224', 'save=False', 'save_period=-1', 'cache=False', 'device=None', 'workers=8', 'project=None', 'name=None', 'exist_ok=False', 'pretrained=True', 'optimizer=auto', 'verbose=True', 'seed=0', 'deterministic=True', 'single_cls=False', 'rect=False', 'cos_lr=False', 'close_mosaic=10', 'resume=False', 'amp=True', 'fraction=1.0', 'profile=False', 'freeze=None', 'multi_scale=False', 'overlap_mask=True', 'mask_ratio=4', 'dropout=0.0', 'val=False', 'split=val', 'save_json=False', 'save_hybrid=False', 'conf=None', 'iou=0.7', 'max_det=300', 'half=False', 'dnn=False', 'plots=False', 'source=None', 'vid_stride=1', 'stream_buffer=False', 'visualize=False', 'augment=False', 'agnostic_nms=False', 'classes=None', 'retina_masks=False', 'embed=None', 'show=False', 'save_frames=False', 'save_txt=False', 'save_conf=False', 'save_crop=False', 'show_labels=True', 'show_conf=True', 'show_boxes=True', 'line_width=None', 'format=torchscript', 'keras=False', 'optimize=False', 'int8=False', 'dynamic=False', 'simplify=False', 'opset=None', 'workspace=4', 'nms=False', 'lr0=0.01', 'lrf=0.01', 'momentum=0.937', 'weight_decay=0.0005', 'warmup_epochs=3.0', 'warmup_momentum=0.8', 'warmup_bias_lr=0.1', 'box=7.5', 'cls=0.5', 'dfl=1.5', 'pose=12.0', 'kobj=1.0', 'label_smoothing=0.0', 'nbs=64', 'hsv_h=0.015', 'hsv_s=0.7', 'hsv_v=0.4', 'degrees=0.0', 'translate=0.1', 'scale=0.5', 'shear=0.0', 'perspective=0.0', 'flipud=0.0', 'fliplr=0.5', 'bgr=0.0', 'mosaic=1.0', 'mixup=0.0', 'copy_paste=0.0', 'auto_augment=randaugment', 'erasing=0.4', 'crop_fraction=1.0', 'cfg=None', 'tracker=botsort.yaml']' returned non-zero exit status 1.

How can i fix it?
Thanks for reading!

**Comment by glenn-jocher** [comment]
It looks like you're encountering an error with the `tune` method. Please ensure you have the latest versions of Ultralytics and Ray Tune installed. If the issue persists, try running the command without optional parameters like `val=False` and `save=False`. If you still face issues, please share the complete error message for further assistance.

**Comment by 7rkMnpl** [comment]
![opti](https://github.com/user-attachments/assets/614413b2-ed55-47d7-a9ce-cb0b34717662)
and i given data_path = '/dataset/data.yaml'
epochs = 50
batch_size = 8
img_size = 512
patience = 5
learning_rate = 0.00001  # Set your custom learning rate here
. its not taking LR but runs by default for training  cud you pls help me to rectify this

**Comment by 7rkMnpl** [comment]
i am using custom traing code not CLI

**Comment by glenn-jocher** [comment]
@7rkMnpl to set a custom learning rate, ensure you're passing it correctly in your training command or script. If you're using the CLI, it should be `yolo train data=data.yaml lr0=0.00001`. If you're using Python, include `lr0=0.00001` in your `train()` method arguments.

**Comment by 7rkMnpl** [comment]
> Ray Tune

ok but still its not taking custom optimizer.

**Comment by glenn-jocher** [comment]
@7rkMnpl to use a custom optimizer, ensure you're specifying it correctly in your training command or script. If you're using the CLI, add `optimizer=Adam` (or your desired optimizer). If using Python, include `optimizer='Adam'` in your `train()` method arguments.

**Comment by Hijaaaz** [comment]
Hi

**Comment by glenn-jocher** [comment]
@Hijaaaz to set a custom optimizer and learning rate in YOLOv8, specify them directly in your training command. For CLI: add `optimizer=Adam lr0=0.00001` to your existing command. For Python: include these parameters in your `model.train()` call. Ensure your Ultralytics version is updated. The full list of train parameters is available in our [configuration docs](https://docs.ultralytics.com/usage/cfg/).
